File: src/apps/EventViewer/src/KeyResponder.m
#import "KeyResponder.h"
#import "Karabiner_EventViewer-Swift.h"

@interface KeyResponder ()

@property(weak) IBOutlet EventQueue* eventQueue;

@end

@implementation KeyResponder

- (void)drawRect:(NSRect)dirtyRect {
  [NSGraphicsContext saveGraphicsState];
  {
    NSRect bounds = [self bounds];

    CGFloat w = bounds.size.width / 40;
    NSRect text = NSMakeRect(w * 2,
                             w * 2,
                             bounds.size.width - w * 4,
                             bounds.size.height - w * 4);

    // Draw area
    [[[NSColor whiteColor] colorWithAlphaComponent:0.8] set];
    [[NSBezierPath bezierPathWithRoundedRect:bounds xRadius:10 yRadius:10] fill];

    [[[NSColor blackColor] colorWithAlphaComponent:0.6] set];
    NSBezierPath* path = [NSBezierPath bezierPathWithRoundedRect:bounds xRadius:10 yRadius:10];
    [path setLineWidth:w];
    [path stroke];

    // Draw texts
    NSDictionary* attributes = @{
      NSFontAttributeName : [NSFont boldSystemFontOfSize:(text.size.width / 10)],
      NSForegroundColorAttributeName : [[NSColor blackColor] colorWithAlphaComponent:0.6],
    };
    [@"Mouse Area" drawInRect:text withAttributes:attributes];
  }
  [NSGraphicsContext restoreGraphicsState];
}

// ============================================================
//
// Key event handlers
//

- (void)keyDown:(NSEvent*)event {
}

- (void)keyUp:(NSEvent*)event {
}

- (void)flagsChanged:(NSEvent*)event {
}

- (void)mouseDown:(NSEvent*)event {
  [self.eventQueue pushMouseEvent:event];
}

- (void)mouseUp:(NSEvent*)event {
  [self.eventQueue pushMouseEvent:event];
}

- (void)mouseDragged:(NSEvent*)event {
  [self.eventQueue pushMouseEvent:event];
}

- (void)rightMouseDown:(NSEvent*)event {
  [self.eventQueue pushMouseEvent:event];
}

- (void)rightMouseUp:(NSEvent*)event {
  [self.eventQueue pushMouseEvent:event];
}

- (void)rightMouseDragged:(NSEvent*)event {
  [self.eventQueue pushMouseEvent:event];
}

- (void)otherMouseDown:(NSEvent*)event {
  [self.eventQueue pushMouseEvent:event];
}

- (void)otherMouseUp:(NSEvent*)event {
  [self.eventQueue pushMouseEvent:event];
}

- (void)otherMouseDragged:(NSEvent*)event {
  [self.eventQueue pushMouseEvent:event];
}

- (void)scrollWheel:(NSEvent*)event {
  [self.eventQueue pushMouseEvent:event];
}

- (void)swipeWithEvent:(NSEvent*)event {
}

@end

File: src/apps/EventViewer/src/KeyResponder.swift
import Cocoa

public class KeyResponder: NSView {
    @IBOutlet var eventQueue: EventQueue!

    override public func draw(_: NSRect) {
        NSGraphicsContext.saveGraphicsState()

        //
        // Draw area
        //

        let lineWidth = bounds.size.width / 40.0

        NSColor.white.withAlphaComponent(0.8).set()
        NSBezierPath(roundedRect: bounds, xRadius: 10.0, yRadius: 10.0).fill()

        NSColor.black.withAlphaComponent(0.6).set()
        let path = NSBezierPath(roundedRect: bounds, xRadius: 10, yRadius: 10)
        path.lineWidth = lineWidth
        path.stroke()

        //
        // Draw texts
        //

        let textRect = NSMakeRect(lineWidth * 2,
                                  lineWidth * 2,
                                  bounds.size.width - lineWidth * 4,
                                  bounds.size.height - lineWidth * 4)

        let attributes: [NSAttributedString.Key: Any] = [
            NSAttributedString.Key.font: NSFont.boldSystemFont(ofSize: textRect.size.width / 10),
            NSAttributedString.Key.foregroundColor: NSColor.black.withAlphaComponent(0.6),
        ]

        NSString("Mouse Area").draw(in: textRect, withAttributes: attributes)

        NSGraphicsContext.restoreGraphicsState()
    }

//
    // Key event handlers
//

    override public func keyDown(with _: NSEvent) {}

    override public func keyUp(with _: NSEvent) {}

    override public func flagsChanged(with _: NSEvent) {}

    override public func mouseDown(with event: NSEvent) {
        eventQueue.pushMouseEvent(event)
    }

    override public func mouseUp(with event: NSEvent) {
        eventQueue.pushMouseEvent(event)
    }

    override public func mouseDragged(with event: NSEvent) {
        eventQueue.pushMouseEvent(event)
    }

    override public func rightMouseDown(with event: NSEvent) {
        eventQueue.pushMouseEvent(event)
    }

    override public func rightMouseUp(with event: NSEvent) {
        eventQueue.pushMouseEvent(event)
    }

    override public func rightMouseDragged(with event: NSEvent) {
        eventQueue.pushMouseEvent(event)
    }

    override public func otherMouseDown(with event: NSEvent) {
        eventQueue.pushMouseEvent(event)
    }

    override public func otherMouseUp(with event: NSEvent) {
        eventQueue.pushMouseEvent(event)
    }

    override public func otherMouseDragged(with event: NSEvent) {
        eventQueue.pushMouseEvent(event)
    }

    override public func scrollWheel(with event: NSEvent) {
        eventQueue.pushMouseEvent(event)
    }

    override public func swipe(with _: NSEvent) {}
}
